Feature/first deliverable

LoliDelgado opened the pull request:
### Description
I am creating a Golang project, having as source data a CSV file, and presenting it as an API.

### Screenshots:
Get all users:
![Screen Shot 2022-06-10 at 15 09 08](https://user-images.githubusercontent.com/81374700/173143753-8388223b-7977-49ed-8ac5-1d84a0aacb48.png)

Get inexistent user:
![Screen Shot 2022-06-10 at 15 09 42](https://user-images.githubusercontent.com/81374700/173143748-5260112b-0afe-4bdb-bda1-f94bd994bcc6.png)

Get existent user:
![Screen Shot 2022-06-10 at 15 09 36](https://user-images.githubusercontent.com/81374700/173143751-97627c07-1939-4493-95fe-cc64f2377607.png)

File not found:
![Screen Shot 2022-06-13 at 11 19 28](https://user-images.githubusercontent.com/81374700/173399203-72b3a920-11ce-493e-8807-43119c8de13f.png)

Bad request:
![Screen Shot 2022-06-13 at 11 20 34](https://user-images.githubusercontent.com/81374700/173399360-5d02dc61-ea36-46e8-b6d8-fc8dbdd7b974.png)

LoliDelgado committed:
feat: :tada: Create endpoint using clean architecture

First iteration to have an API whose source is a CVS file, delivering data as JSON, trying to structure the project based on the clean architecture methodology

LoliDelgado committed:
refactor: :recycle: Move logic to read file to repository layer

Instead of having to open the file from main, I am moving the logic to repository, and from main I am only sending the name of the file to be read.

LoliDelgado committed:
docs: :memo: Updating comments on usecase file

LoliDelgado committed:
test: :white_check_mark: Add test for githubuser model

Created function to have a githubUser, and added its test, after passing it using the function in the repository

LoliDelgado committed:
feat: :sparkles: Add endpoint to filter by id

A new endpoint to filter the users by id has been added, also for that I have changed to use gorilla/mux package to handle the requests

\

LoliDelgado committed:
refactor: :children_crossing: Add reponse codes

Handling errors now returning the needed status code and providing info about errors when they happen, also, addes the ok status when the reponse is correct

LoliDelgado committed:
refactor: :children_crossing: Add reponse codes

Handling errors now returning the needed status code and providing info about errors when they happen, also, addes the ok status when the reponse is correct

RolloCasanova commented on go.mod:
```diff
@@ -0,0 +1,11 @@
+module lolidelgado/github-users
```
Just a recommendation; normally the module is named after the github repo path.

This is done this way because maybe in a future someone decides to import your project, and as the name is `lolidelgado/github-users`, then it will lead to a failure to fetch this route.

In this case, the name should be:

```suggestion
module github.com/lolidelgado/ondemand-go-bootcamp
```

LoliDelgado replied:
Module name updated, thanks for the recommendation!

RolloCasanova commented on main.go:
```diff
@@ -0,0 +1,35 @@
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/unrolled/render"
```
A good practice in Go is to separate all your imports into three main blocks:
- 1st - Standard library packages
- 2nd - Internal (project) packages
- 3rd - Third party packages

You can swap 2nd and 3rd blocks, and also each block should contain all of its elements in alphabetical order.

```suggestion
	"log"
	"net/http"
	
	"lolidelgado/github-users/controller"
	"lolidelgado/github-users/delivery"
	"lolidelgado/github-users/repository"
	"lolidelgado/github-users/usecase"

	"github.com/gorilla/mux"
	"github.com/unrolled/render"
```

Please fix this and all other occurrences of this package blocking issue.

LoliDelgado replied:
Ordering and separating the packages following that good practice, in fact, I think when saving the files the imports were automatically ordered, maybe I missed something the first time I created them. Thanks!

RolloCasanova commented on main.go:
```diff
@@ -0,0 +1,35 @@
+		httpRouter,
+	)
+
+	//start server
```
Just nitpicking here, I think leaving a space after `//` gives better readability.
```suggestion
	// repository
	githubUserRepo := repository.NewGithubUser(fileName)

	// useCase
	githubUserUseCase := usecase.NewGithubUser(githubUserRepo)

	// controllers for Rest
	httpRender := render.New()
	usersController := controller.NewGithubUser(httpRender, githubUserUseCase)

	httpRouter := mux.NewRouter()
	delivery.Setup(
		usersController,
		httpRouter,
	)

	// start server
```

RolloCasanova commented on main.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+func main() {
+	const fileName = "github_users.csv"
```
Normally constants live on the package level.

Also, to take it into consideration, let's avoid having magic naming(hardcoded values) in the code. I know you require this value to initialize the repo, but there are packages that allow us read values from certain places, such as a config file, an .env file, from an env var, etc.

In a scenario in which you decide to change the filename, you will end creating a new commit; if we have this value dynamic, there's no need to make changes in our code base.

I suggest you to give it a look to [viper](https://github.com/spf13/viper).

LoliDelgado replied:
Thanks for the suggestion! updated the code and created a config file just to handle all this kind of data that might change easily

RolloCasanova commented on main.go:
```diff
@@ -0,0 +1,35 @@
+	)
+
+	//start server
+	log.Fatal(http.ListenAndServe(":7000", httpRouter))
```
I'd suggest to add a hint to the end user that the service is running.
```suggestion
	log.Info("Starting server at port 7000...")
	log.Fatal(http.ListenAndServe(":7000", httpRouter))
```

RolloCasanova replied:
Also, regarding the magic naming and magic numbers, in this case I'd also recommend to make the port a dynamic value; what if that port is already in use? you need to change your code base to amend that.

If you read that value from an external source you won't need that change in the code base.

LoliDelgado replied:
Updated, now the port is stored in an environment variable and can be updated if needed

RolloCasanova commented on delivery/router.go:
```diff
@@ -0,0 +1,13 @@
+
+	router.HandleFunc("/github-users", githubUserHandler.GetGithubUsers)
+	router.HandleFunc("/github-users/{id}", githubUserHandler.GetGithubUserById)
+}
```
A pro-tip, mux contains a method named `Methods` that allows you to specify which HTTP methods you are going to support on your `handleFunc` for the given route. No need to validate in the controller if that's a GET request.

Also, it is important to always comment your functions, methods, structs, variables, constants, etc., as for you it makes sense what it does, but for another person that might want to import and use your package, it gives an idea what this element does, and also it works as a way to document your code; this is the standard in Go.

You can give it a look into https://go.dev/doc/comment for a more in-deep explanation. 
```suggestion
// Setup adds the routes to the given router
func Setup(githubUserHandler *controller.GithubUser, router *mux.Router) {
	router.HandleFunc("/github-users", githubUserHandler.GetGithubUsers).Methods(http.MethodGet)
	router.HandleFunc("/github-users/{id}", githubUserHandler.GetGithubUserById).Methods(http.MethodGet)
}
```

RolloCasanova commented on controller/github_user.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/unrolled/render"
+)
+
+type GithubUser struct {
```
Don't forget to add a comment for all funcs, structs, methods, etc.

RolloCasanova commented on controller/github_user.go:
```diff
@@ -0,0 +1,49 @@
+		c.render.JSON(rw, http.StatusOK, githubUsers)
+	}
+
+}
```
As the comment in the `router.go file`, you can now get rid of this switch, as you already know this is a method specific for get request and it's already filtered at router level.

```suggestion
// GetGithubUser fetches all github users from the usecase
func (c *GithubUser) GetGithubUsers(rw http.ResponseWriter, req *http.Request) {
	githubUsers, err := c.g.FetchAll()
	if err != nil {
		c.render.Text(rw, http.StatusInternalServerError, fmt.Sprintf("This is not fine🔥\nAnd the reason is: %s", err))
		return
	}
	
	c.render.JSON(rw, http.StatusOK, githubUsers)
}
```

RolloCasanova commented on controller/github_user.go:
```diff
@@ -0,0 +1,49 @@
+	case http.MethodGet:
+		githubUsers, err := c.g.FetchAll()
+		if err != nil {
+			c.render.Text(rw, http.StatusInternalServerError, fmt.Sprintf("This is not fine🔥\nAnd the reason is: %s", err))
```
Cool usage of emojis!

RolloCasanova commented on controller/github_user.go:
```diff
@@ -0,0 +1,49 @@
+		c.render.Text(rw, http.StatusInternalServerError, fmt.Sprintf("This is not fine🔥\nAnd the reason is: %s", err))
+		return
+	}
+	c.render.JSON(rw, http.StatusOK, githubUser)
```
Nitpicking here, it's better to add here and there some empty lines for better readability, and separation of logic steps.

```suggestion
	vars := mux.Vars(req)
	id := vars["id"]
	
	numericId, err := strconv.Atoi(id)
	if err != nil {
	 	c.render.Text(rw, http.StatusBadRequest, "invalid id")
	 	
		return
	}
	
	githubUser, err := c.g.GetById(numericId)
	if err != nil {
		c.render.Text(rw, http.StatusInternalServerError, fmt.Sprintf("This is not fine🔥\nAnd the reason is: %s", err))
		
		return
	}
	
	c.render.JSON(rw, http.StatusOK, githubUser)
```

RolloCasanova commented on models/github_user.go:
```diff
@@ -0,0 +1,14 @@
+package models
```
Nitpicking here, the good practice in Go is to name the packages in singular.
```suggestion
package model
```

LoliDelgado replied:
thanks for the tip! updated it

RolloCasanova commented on models/github_user.go:
```diff
@@ -0,0 +1,14 @@
+package models
+
+type GithubUser struct {
+	Id          int
```
Another not written rule is to write acronyms in all letters upper case.
```suggestion
	ID          int
```

RolloCasanova commented on models/github_user.go:
```diff
@@ -0,0 +1,14 @@
+	PublicRepos int
+}
+
+func NewGithubUser(id int, login, name, company, bio string, publicrepos int) GithubUser {
```
Nitpicking here, I've seen constructors on models, but in practice, and the way to read code in Go, it's way easier to create a variable as following:

```code
user := model.GithubUser{
	ID: 1,
	Login: "RolloCasanova",
	Company: "Wizeline",
}
```

This way, in case you don't have any of the values in the params you will avoid sending empty strings. Remember, when initializing a variable as the example I provided, any non-given value will be assigned the zero value.

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+
+type IGithubUserRepository interface {
+	FetchAll() ([]models.GithubUser, error)
+}
```
Excellent, using interfaces is a best practice in clean architecture. I'd also recommend to use them in the controller and the usecase.

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+	FetchAll() ([]models.GithubUser, error)
+}
+
+const staticFilesPath = "static/"
```
IMHO, this should be a field of your `NewGithubUser` struct, which needs to be sent from the main function, and also be a dynamic value.

Same reason as the fileName, if later you decide to change the location of the file, your codebase should remain the same; the only thing that would need to change is the value from config.

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+	return arrayToGithubUser(lines), nil
+}
+
+func (g *GithubUser) GetById(id int) ([]models.GithubUser, error) {
```
Question, if you are getting a single GithubUser, why your return here is an array?

As mentioned before, this should be named `GetByID`.

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+			break
+		}
+	}
+	if len(user) > 0 {
+		return arrayToGithubUser(user), nil
+	} else {
+		return []models.GithubUser{}, nil
+	}
```
Pro-tip, you can get rid off the else statement as following:

```suggestion
	if len(user) > 0 {
		return arrayToGithubUser(user), nil
	}
	
	return []models.GithubUser{}, nil
```

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+			break
+		}
+		if err != nil {
+			log.Fatal(err)
```
You should not call log.Fatal here.

Normally you back-propagate the error to the caller of this function; in this case it was in the same package, but also that function was called on the usecase, which was called by the controller.

Also, if you call `log.Fatal` here, the defer in line 62 will never be called.

Controller should be the one that handles this error; and avoiding ending the service execution abruptly.

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func arrayToGithubUser(lines [][]string) []models.GithubUser {
+	var githubUsers []models.GithubUser
```
Consider preallocate the githubUsers array. if you already have an idea of how many elements you have, it's better to initialize the array with that length (or capacity).

RolloCasanova commented on repository/github_user.go:
```diff
@@ -0,0 +1,96 @@
+	return lines, nil
+}
+
+func arrayToGithubUser(lines [][]string) []models.GithubUser {
+	var githubUsers []models.GithubUser
+	for _, line := range lines {
+		id, err := strconv.Atoi(line[0])
+		if err != nil {
+			continue
+		}
+		publicRepos, err := strconv.Atoi(line[5])
+		if err != nil {
+			publicRepos = 0
+		}
+		data := models.NewGithubUser(id, line[1], line[2], line[3], line[4], publicRepos)
+		githubUsers = append(githubUsers, data)
+	}
+	return githubUsers
+}
```
Nitpicking here, add empty lines for better readability.
```suggestion 
func arrayToGithubUser(lines [][]string) []models.GithubUser {
	var githubUsers []models.GithubUser
	
	for _, line := range lines {
		id, err := strconv.Atoi(line[0])
		if err != nil {
			continue
		}
		
		publicRepos, err := strconv.Atoi(line[5])
		if err != nil {
			publicRepos = 0
		}
		
		data := models.NewGithubUser(id, line[1], line[2], line[3], line[4], publicRepos)
		githubUsers = append(githubUsers, data)
	}
	
	return githubUsers
}
```

RolloCasanova commented on usecase/github_user.go:
```diff
@@ -0,0 +1,25 @@
+//Package usecase processes the information from repository, making any needed calculations, data transformation, etc.
+//If it recieved data it's from controllers, coming from delivery layer
+package usecase
```
Cool usage of package scope comments.

A little typo:

```suggestion
//Package usecase processes the information from repository, making any needed calculations, data transformation, etc.
//If it received data it's from controllers, coming from delivery layer
package usecase
```

RolloCasanova commented on usecase/github_user.go:
```diff
@@ -0,0 +1,25 @@
+	}
+}
+
+func (g *GithubUserUseCase) FetchAll() ([]models.GithubUser, error) {
+	return g.githubUserRepo.FetchAll()
+}
```
Something important in Go at the time of determining the root cause of an error is to wrap errros.

Let's say you have more logic here, and you simply decide to return what you get from the repository, and also let's imagine you need to determine the reason of this error. You won't know at glance if the repo package, or the usecase package, or any other package was the pace  where the error was originated.

By wrapping the error in each layer you are adding useful information that will help debugging tasks.

To wrap errors, there's a method named `Errorf` in the `fmt`pakage.

```suggestion
func (g *GithubUserUseCase) FetchAll() ([]models.GithubUser, error) {
	githubUsers, err := g.githubUserRepo.FetchAll()
	if err != nil {
		return nil, fmt.Errorf("Error fetching github users: %w", err)
	}

	return githubUsers, nil
}

```

LoliDelgado replied:
Is it better if I handle the errors from the bottom, indicating errors and reasons from the repository, I am going for that approach but please let me know if there is another way that is more suggested as best practice.

RolloCasanova replied:
Well, the most common way I've seen in my experience is to back-propagate the error in each layer and adding information in each one of the layers; there is not a best practice that reigns above all, but reading [this](https://earthly.dev/blog/golang-errors/ ) might come handy to take a decision.

- You can return an error created on the fly, just as the suggestion I added.
- You can initialize it as a package-level error variable and use it multiple times.
- You can define your own errors package, with error definitions, methods, and functions for the whole project.

Any of the choice you take, remember that you can also define your own errors; `error` is an interface that consists on only an `Error() string` method.

Choose the approach that better suits your needs; remember that part of the deliverable is to include unit tests for your code; maybe an already defined error works better, but the approach is according to how exactly you want to organize your project

And remember, it's always a good idea to wrap errors.

RolloCasanova reviewed (changes requested):
That was a very strong first approach. You have been started using clean architecture pretty well.

I added a bunch of comments, but more of them are regarding some not written rules in Go, and some best practices.

Something I'd like to add is regarding the README.md. file. I know you added it in the first PR you submitted, but it would look good in the README.md as well. It would be good to see some instructions or usage examples for people who might not be related or know the project's specs. You can move the actual README.md to another file, or delete it, and reuse it for these purposes.

Also, there's some issues I found when running the program. One was the port 7000 was somehow already in use in my machine, so I had to change it to a common one (3000, 8000). Also, when creating a POST request to an endpoint, I'm getting a 200 OK response, when that shouldn't be.

Another thing, when getting a single GithubUser that does not exists (ID 1), I get an empty array as a response and also a 200 OK status; be careful with the response type and the status you return.

And lastly, when getting a single GithubUser, and it does exists, I get the response with the info of that user, but the response type is an array, where IMHO should be a simple object.

Any doubt you might have don't hesitate to ask me here in the PR or via Slack DM; even you can setup a 1:1 meeting with me or any other of the mentors of the bootcamp.

Will be pending of your response and future PR deliveries.

LoliDelgado committed:
refactor: :recycle: Rename modules

Following best practices to name the modules and ordering correctly the packages when importing them

LoliDelgado committed:
refactor: :lipstick: Add endpoint method in the route itself

Handling the method accepted from the delivery layer instead of doing it in the controller

LoliDelgado committed:
refactor: Update readme and little updates on code

LoliDelgado committed:
refactor: :recycle: Handle errors and adding interfaces

Also adding some break lines for more readability

LoliDelgado committed:
feat: :sparkles: Add viper to handle configurations

LoliDelgado committed:
refactor: :recycle: Handling errors and wraping them

Addes some structs to handle better the errors, which are now defined as Custom Errors

LoliDelgado committed:
refactor: :recycle: Clean errors and delete constructor from model

Now using the model to create instances directly of it without having a constructor, for that I am deleting the test case for the same. Also updated the errors defined from repository for those which are really useful
